pkg/controller/job/job_controller.go

```diff
@@ -628,37 +631,60 @@ func (jm *Controller) syncOrphanPod(ctx context.Context, key string) error {
 	}()
 
 	ns, name, err := cache.SplitMetaNamespaceKey(key)
+
+	if err == nil {
+		// If splitting is successful, treat the key as a pod name.
+		return jm.handleSingleOrphanPod(ctx, ns, name)
+	}
+
+	// If there's an error splitting, assume the key is a label selector
+	logger.V(4).Info("Key is assumed to be a label selector, syncing all matching pods", "labelSelector", key)
+	return jm.syncOrphanPodsBySelector(ctx, key)
+}
+
+// syncOrphanPodsBySelector fetches and processes all pods matching the given label selector.
+func (jm *Controller) syncOrphanPodsBySelector(ctx context.Context, labelSelector string) error {
+	logger := klog.FromContext(ctx)
+	selector, err := labels.Parse(labelSelector)
+	if err != nil {
+		return fmt.Errorf("invalid label selector: %w", err)
+	}
+
+	// Fetch all pods that match the label selector.
+	// relatively expensive operation but given it is in reconciler(not in sync path) and that too as part of orphan reconciler it is ok
+	pods, err := jm.podStore.Pods("").List(selector)
 	if err != nil {
 		return err
 	}
+	for _, pod := range pods {
+		if err := jm.handleSingleOrphanPod(ctx, pod.Namespace, pod.Name); err != nil {
+			logger.Error(err, "syncing orphan pod failed", "pod", pod.Name)
+		}
+	}
+	return nil
+}
 
+// handleSingleOrphanPod processes a single orphan pod (logic remains unchanged).
+func (jm *Controller) handleSingleOrphanPod(ctx context.Context, ns, name string) error {
+	logger := klog.FromContext(ctx)
 	sharedPod, err := jm.podStore.Pods(ns).Get(name)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			logger.V(4).Info("Orphan pod has been deleted", "pod", key)
+			logger.V(4).Info("Orphan pod has been deleted", "pod", name)
 			return nil
 		}
 		return err
 	}
-	// Make sure the pod is still orphaned.
 	if controllerRef := metav1.GetControllerOf(sharedPod); controllerRef != nil {
 		if controllerRef.Kind != controllerKind.Kind || controllerRef.APIVersion != batch.SchemeGroupVersion.String() {
-			// The pod is controlled by an owner that is not a batch/v1 Job. Do not remove finalizer.
 			return nil
 		}
 		job := jm.resolveControllerRef(sharedPod.Namespace, controllerRef)
-		if job != nil {
-			// Skip cleanup of finalizers for pods owned by a job managed by an external controller
-			if controllerName := managedByExternalController(job); controllerName != nil {
-				logger.V(2).Info("Skip cleanup of the job finalizer for a pod owned by a job that is managed by an external controller", "key", key, "podUID", sharedPod.UID, "jobUID", job.UID, "controllerName", controllerName)
-				return nil
-			}
-		}
-		if job != nil && !util.IsJobFinished(job) {
-			// The pod was adopted. Do not remove finalizer.
+		if job != nil && (managedByExternalController(job) != nil || !util.IsJobFinished(job)) {
```

Comment:
Are these changes required? If not, please revert.

Reply:
Basically cleaned up this section. Simplified it, not related to performance improvement.

Reply:
Please revert it then. If you have an idea for a cleanup then feel free to open a dedicated PR, but I think we should keep the comment `logger.V(2).Info("Skip cleanup` which might be relevant in some scenarios.